### PR TITLE
IDCs: Add option to write IDCDelta to calibration file

### PR DESCRIPTION
**Pull request by matthias-kleiner**

### CODE REVIEW

File: Detectors/TPC/workflow/include/TPCWorkflow/TPCFactorizeIDCSpec.h [hunk 1]
```diff
@@ -325,9 +327,24 @@ class TPCFactorizeIDCSpec : public o2::framework::Task
...
+              LOGP(info, "Sending object {} / {} of size {} bytes, valid for {} : {} ", ccdbInfoIDCDelta.getPath(), ccdbInfoIDCDelta.getFileName(), imageIDCDelta->size(), ccdbInfoIDCDelta.getStartValidityTimestamp(), ccdbInfoIDCDelta.getEndValidityTimestamp());
+              output.snapshot(Output{o2::calibration::Utils::gDataOriginCDBPayload, getDataDescriptionCCDBIDCDelta(), iChunk}, *imageIDCDelta.get());
+              output.snapshot(Output{o2::calibration::Utils::gDataOriginCDBWrapper, getDataDescriptionCCDBIDCDelta(), iChunk}, ccdbInfoIDCDelta);
+            }
+
+            if (mDumpIDCDeltaCalibData) {
+              const std::string sideStr = sideA ? "A" : "C";
+              TFile fCalibData(fmt::format("IDCDelta_side{}_cal_data_{}.root", sideStr, ccdbInfoIDCDelta.getStartValidityTimestamp()).data(), "RECREATE");
+              fCalibData.WriteObject(imageIDCDelta.get(), "ccdb_object");
+              fCalibData.Close();
```
Comment (shahor02) [inline Detectors/TPC/workflow/include/TPCWorkflow/TPCFactorizeIDCSpec.h]: The imageIDCDelta is already a TFile (memory based), one cannot write it to the TFile, instead one should write it as a binary data:
```
if (mDumpIDCDeltaCalibData && mCalibFileDir != "/dev/null") {
  std::string calibFName = fmt::format("{}IDCDelta_side{}_cal_data_{}.root", mCalibFileDir, sideStr,  ccdbInfoIDCDelta.getStartValidityTimestamp());
try {
 std::ofstream calFile (calibFName, ios::out | ios::binary);
 calFile.write(imageIDCDelta->data(), imageIDCDelta->size());
 calFile.close();
} catch (std::exception const& e) {
  LOG(error) << "Failed to store IDC calibration data file " << calibFName << ", reason: " << e.what();
}
```

File: Detectors/TPC/workflow/include/TPCWorkflow/TPCFactorizeIDCSpec.h [hunk 2]
```diff
@@ -325,9 +327,24 @@ class TPCFactorizeIDCSpec : public o2::framework::Task
...
+
+              TFile fCalibMetaData(fmt::format("IDCDelta_side{}_cal_metadata_{}.root", sideStr, ccdbInfoIDCDelta.getStartValidityTimestamp()).data(), "RECREATE");
+              fCalibMetaData.WriteObject(&ccdbInfoIDCDelta, "ccdb_object");
+              fCalibMetaData.Close();
```
Comment (shahor02) [inline Detectors/TPC/workflow/include/TPCWorkflow/TPCFactorizeIDCSpec.h]: MetaData is a text file, should be created as 
```
if (mMetaFileDir != "/dev/null") {
  o2::dataformats::FileMetaData calMetaData;
  calMetaData.fillFileData(calibFName);
  calMetaData.setDataTakingContext(mDataTakingContext);
  calMetaData.type = "calib;
  calMetaData.priority = "low";
  auto metaFileNameTmp = fmt::format("{}{}.tmp", mMetaFileDir, calibFName);
  auto metaFileName = fmt::format("{}{}.done", mMetaFileDir, calibFName);
  try {
      std::ofstream metaFileOut(metaFileNameTmp);
      metaFileOut << ctfMetaData;
      metaFileOut.close();
      std::filesystem::rename(metaFileNameTmp, metaFileName);
  } catch (std::exception const& e) {
     LOG(error) << "Failed to store CTF meta data file " << metaFileName << ", reason: " << e.what();
  }
}
```
One would need to add the following:
1) fetching mDataTakingContext: add at https://github.com/AliceO2Group/AliceO2/blob/346100773393e2ec822a6848402b92e9a96ca990/Detectors/TPC/workflow/include/TPCWorkflow/TPCFactorizeIDCSpec.h#L191
```
o2::framework::DataTakingContext mDataTakingContext{};
```
and at https://github.com/AliceO2Group/AliceO2/blob/346100773393e2ec822a6848402b92e9a96ca990/Detectors/TPC/workflow/include/TPCWorkflow/TPCFactorizeIDCSpec.h#L89:
```
mDataTakingContext = pc.services().get<DataTakingContext>();
```

2) passing output directories for calib data and metafile: 
add at https://github.com/AliceO2Group/AliceO2/blob/346100773393e2ec822a6848402b92e9a96ca990/Detectors/TPC/workflow/include/TPCWorkflow/TPCFactorizeIDCSpec.h#L169 data members
```
std::string mMetaFileDir{};
std::string mCalibFileDir{};
```
add at https://github.com/AliceO2Group/AliceO2/blob/346100773393e2ec822a6848402b92e9a96ca990/Detectors/TPC/workflow/include/TPCWorkflow/TPCFactorizeIDCSpec.h#L429
```
            {"output-dir", VariantType::String, "none", {"calibration files output directory, must exist"}},
            {"meta-output-dir", VariantType::String, "/dev/null", {"calibration metadata output directory, must exist (if not /dev/null)"}},
```
and at https://github.com/AliceO2Group/AliceO2/blob/346100773393e2ec822a6848402b92e9a96ca990/Detectors/TPC/workflow/include/TPCWorkflow/TPCFactorizeIDCSpec.h#L65
```
mCalibFileDir = ic.options().get<std::string>("output-dir");
if (mCalibFileDir != "/dev/null") {
  mCalibFileDir = o2::utils::Str::rectifyDirectory(mCalibFileDir);
}
mMetaFileDir = ic.options().get<std::string>("meta-output-dir");
if (mMetaFileDir != "/dev/null") {
  mMetaFileDir = o2::utils::Str::rectifyDirectory(mMetaFileDir);
}
```

**Review by shahor02 — CHANGES_REQUESTED** [review]
Hi @matthias-kleiner 
Some fixes are needed, see below.

**Comment by matthias-kleiner** [comment]
> Hi @matthias-kleiner Some fixes are needed, see below.

Hi @shahor02 
many thanks for your detailed help. I implemented the changes and tested it locally. It creates several files now like:
```
IDCDelta_sideA_cal_data_1668222565120.root
.ls
KEY: o2::tpc::IDCDelta<unsigned char>	ccdb_object;1	object title

IDCDelta_sideA_cal_data_1668222565120.root.done
LHCPeriod: NOV
run: 0
lurl: /data/mkleiner/tmp/IDCDelta_sideA_cal_data_1668222565120.root
type: calib
ctime: 1668270541446
size: 8385536
md5: 74f623f10e86acf026b72e6da1e3cfb4
priority: low
det_composition: unknown
```

So I assume it is working.

**Review by shahor02 — DISMISSED** [review]
Thanks, looks fine.
In the real workflow on the EPN these files will go to dedicated directories. 
@ole, where do you set the directories for residual files? I don't see `--output-dir` and `--meta-output-dir` set there.

**Review by shahor02 — APPROVED** [review]

**Pull request merged**